src/big/big_int.cr

```diff
@@ -115,6 +115,78 @@ struct BigInt < Int
     end
   end
 
+  # a.add(b), a += b, avoiding allocations by mutating a
+  #
+  #    a = 100.to_big_i
+  #    a.add(10) => # a = 110
+  def add(other : BigInt) : self
+    LibGMP.add(self, self, other)
+    self
+  end
+
+  # a.add(b), a += b, avoiding allocations by mutating a
+  def add(other : Int) : self
+    if other < 0
+      sub(other.abs)
+    elsif other <= LibGMP::ULong::MAX
+      LibGMP.add_ui(self, self, other)
+    else
+      add(other.to_big_i)
+    end
+    self
+  end
+
+  # a.add(b,c), a = b+c avoiding allocations by mutating a. This is a nice way
+  # to "reuse" `BigInt` variables and avoid new allocations. Beware that
+  # the value of "a" will be overwritten with the result of b*c
+  #
+  #    a = 0.to_big_i
+  #    a.add(10_to_big_i, 5.to_big_i) => # a = 15
+  #
+  #    a = 100.to_big_i
+  #    a.add(10.to_big_i, 5.to_big_i) => # a = 15
+  def add(op1 : BigInt, op2 : BigInt) : self
```

Comment:
I suggest using another name like `save_add` to make it explicit that you're not adding anything to `self` but that you are saving the addition of 2 other numbers to `self`

Reply:
I have no problem with that.

I just wonder if splitting them was necessary since a add taking two arguments normally add them together and return a result, but if you only have 1 argument the only logical thing that can happen is that it's added to the calling object.

If we divide them in two we could also rename `a.add(b)` to `a.add_with(b)` so it's clear that a is added with b, but `save_add` is good too. Do you know if there's any other places in std lib that has to tackle this problem and has a naming convention we could use? I can't think of anything myself.

Reply:
Nothing like this in the stdlib iirc.

FWIW, when I saw `a.add(b, c)` I thought at first that it'd do `a = a + b + c`, not `a = b + c`.

Reply:
OK, good point. I'll see if someone else chimes in with opinions, but I see we might need to change this.

Reply:
or maybe even `add(*ops : BigInt) : self`

Reply:
That being said, I also wouldn't use the method name `add` because it doesn't properly convey the exact meaning of this operation.
I've looked at the Ruby GMP bindings in the [gmp gem](https://www.rubydoc.info/gems/gmp/0.7.43/GMP/Z) where there is a class method `.add(result, op1, op2)` and an instance method `#add!(b)`.
That's not ideal either because it separates both variants (one and two arguments) into completely different methods, yet they implement the same behaviour (just with a default argument). Sill, from a naming perspective, this is much better because `#add!` and `.add` clearly communicate that they are different from say `#+`.

Reply:
Yes I know, I thought about that, but also I thought when someone reads the api it will be clear there are two different uses, and defaulting to self doesn't work for sub so I thought I'd better make them consistent. But yes, no problem to change that.

Reply:
@Sija I don't think it makes any sense to have a splat of operator arguments. And I fail to see any benefit over using a aggregation method like `ops.each { |op| sum.add!(op) }`

Reply:
> That's not ideal either because it separates both variants (one and two arguments) into completely different methods, yet they implement the same behaviour (just with a default argument).

Current implementation is better suited for `.add(result, op1, op2)`. `#add` without `self` is quite misleading.

IMO `#add(!)` implies using `self` as part of the addition - not as an argument (default or not - it's an implementation detail).

Reply:
I looked at the Ruby code too, and didn't get much wiser. As I see it there are three good options:
1. Keep as is and document it
2. Change the `a.add(b, c)` to something else
3. Consider if we really need `a.add(b)`, `a.add(b, c)` is the more flexible of them I think since you can do `a.add(a, b)` if I'm not mistaken.

For 2. could something like `a.add_two(b, c)` work?

Reply:
Go uses `c.Add(a, b)` where the sum a + b and stores the result in c

Reply:
I'd model the API after Ruby, since IMO it's clear about how these operations work, i.e.:

- `#add!(other)` => `a = a + other`
- `.add(result, op1, op2)` => `result = op1 + op2`

Reply:
I agree for `#add!` but not for the second, it's unclear to me that it won't add all args and return a new instance with the result.